File: features/step-definitions/common-steps.js
import Zombie from 'zombie'

import { elementsWithContent, firstTextInput, nthElement, pressEnter } from '../support/utils'

module.exports = function () {
  this.Given(/^I am on the "([^"]+)" page$/, function (page) {
    this.browser = new Zombie()

    return this.browser.visit(`http://localhost:8888/${page}.html`)
  })

  this.When(/^I click the "([^"]+)" link$/, function (text) {
    const link = elementsWithContent(this.browser, 'a', text)[0]

    return this.browser.click(link)
  })

  this.When(/^I press the "([^"]+)" button$/, function (button) {
    return this.browser.pressButton(button)
  })

  this.When(/^I press the "([^"]+)" button (\d+) times?$/, function (button, times) {
    const n = Number(times)
    const ps = Array(n).fill(button).map(b => this.browser.pressButton(b))

    return Promise.all(ps)
  })

  this.When(/^I press the (\d+)(?:st|nd|rd|th) "([^"]+)" button$/, function (nth, className) {
    return this.browser.pressButton(nthElement(this.browser, nth, className))
  })

  this.When(/^I press the (\d+)(?:st|nd|rd|th) "([^"]+)" button (\d+) times?$/, function (nth, className, times) {
    const button = nthElement(this.browser, nth, className)
    const n = Number(times)
    const ps = Array(n).fill(button).map(b => this.browser.pressButton(b))

    return Promise.all(ps)
  })

  this.When(/^I write "([^"]+)" into the input$/, function (val) {
    return this.browser.fill(firstTextInput(this.browser), val)
  })

  this.When(/^I hit enter$/, function (done) {
    const input = firstTextInput(this.browser)

    pressEnter(this.browser, input)

    done()
  })
}

File: features/step_definitions/common_steps.rb
Given(/^I am on the "([^"]*)" page$/) do |page|
  visit "#{page}.html"
end

When(/^I press the "([^"]*)" button$/) do |button|
  click_button button
end

When(/^I click the "([^"]*)" link$/) do |link|
  click_link link
end

When(/^I press the "([^"]*)" button (\d+) times?$/) do |button, n|
  n.to_i.times { click_button button }
end

When(/^I press the (\d+)(?:st|nd|rd|th) "([^"]*)" button$/) do |nth, className|
  click_nth_button(className, nth)
end

When(/^I press the (\d+)(?:st|nd|rd|th) "([^"]*)" button (\d+) times?$/) do |nth, className, times|
  times.to_i.times { click_nth_button(className, nth) }
end

When(/^I write "([^"]*)" into the input$/) do |value|
  first_input.set(value)
end

When(/^I hit enter$/) do
  first_input.native.send_keys(:return)
end
